### ISSUE TITLE
CQL-to-ELM date range optimization doesn't work for Choice Types nor Dates

### ISSUE THREAD

**Issue by ntlanglois** [cqframework/clinical_quality_language#486]
The CQL-to-ELM date range optimization works for Period type, but only if it's not a Choice Type.  It doesn't work for Date type, seemingly except for a strange workaround of introducing an additional CQL define expression occurring prior to the target CQL define expression where both expressions use the same FHIR Resource.  I'm not sure if these issues are overlapping or distinct, but I'm open to splitting into 2 issues if needed.

For context, I am setting the date range optimization option, and it is currently working (returning dateRange and dateProperty) for examples where I use non-ChoiceType Period (e.g. [Encounter.period](https://www.hl7.org/fhir/STU3/encounter-definitions.html#Encounter.period)).  I have not applied the enable interval promotion option, but have tried a few times and saw no difference to these issues.

Example CQL demonstrating issue:
```
library Query version '1'

using FHIR version '3.0.0'

include FHIRHelpers version '3.0.0' called FHIRHelpers

parameter MeasurementPeriod default Interval[DateTime(2014, 1, 1, 0, 0, 0), DateTime(2014, 12, 31, 23, 59, 59))

context Patient

define DateCondition: [Condition] C where C.assertedDate during MeasurementPeriod
define ChoiceTypePeriodCondition: [Condition] C where C.onset as FHIR.Period during MeasurementPeriod
```

Returned ELM contains neither dateRange nor dateProperty, and this remains true even when I remove either one of the defines.

When I simply switch the order of the defines (ChoiceTypePeriodCondition 1st, then DateCondition 2nd), then the returned ELM contains dateRange and dateProperty, but only for DateCondition.

I expected the original example to produce ELM with 2 dateRange and 2 dateProperty occurrences; 1 for each define.  I also expected that each define would produce its corresponding dateRange and dateProperty in the ELM regardless of the presence of other defines.

Thanks for reading, and please let me know if additional information is needed.

**Comment by brynrhodes** [comment]
@ntlanglois , thank you for the detailed report, it's likely that the date range optimization logic isn't correctly accounting for the choice type case, we'll take a look and get this addressed.

**Comment by brynrhodes** [comment]
Addressed in 1.4.9-SNAPSHOT

**Comment by brynrhodes** [comment]
Addressed in 1.4.9 release
